osquery/remote/transports/tests/tls_transports_tests.cpp

```diff
@@ -0,0 +1,216 @@
+
+#include "osquery/core/test_util.h"
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+
+class TLSServerRunner : private boost::noncopyable {
+ public:
+  /// Create a singleton TLS server runner.
+  static TLSServerRunner& instance() {
+    static TLSServerRunner instance;
+    return instance;
+  }
+
+  /// TCP port accessor.
+  static const std::string& port() { return instance().port_; }
+  /// Start the server if it hasn't started already.
+  static void start();
+  /// Stop the service when the process exits.
+  static void stop();
+
+ private:
+  TLSServerRunner()
+      : server_(0), port_(std::to_string(rand() % 10000 + 20000)){};
+  TLSServerRunner(TLSServerRunner const&);
+  void operator=(TLSServerRunner const&);
+  virtual ~TLSServerRunner() { stop(); }
+
+ private:
+  pid_t server_;
+  std::string port_;
+};
+
+void TLSServerRunner::start() {
+  auto& self = instance();
+  if (self.server_ != 0) {
+    return;
+  }
+
+  // Fork then exec a shell.
+  self.server_ = fork();
+  if (self.server_ == 0) {
+    // Start a python TLS/HTTPS or HTTP server.
+    auto script = kTestDataPath + "/test_http_server.py --tls " + self.port_;
+    execlp("sh", "sh", "-c", script.c_str(), nullptr);
+    ::exit(0);
+  }
+  ::sleep(1);
+}
+
+void TLSServerRunner::stop() {
+  auto& self = instance();
+  kill(self.server_, SIGTERM);
+}
+
+class TLSTransportsTests : public testing::Test {
+ public:
+  bool verify(const Status &status) {
+    if (!status.ok()) {
+      LOG(ERROR) << "Could not complete TLSRequest (" << status.getCode()
+                 << "): " << status.what();
+    }
+
+    // Sometimes the best we can test is the call workflow.
+    if (status.getCode() == 1) {
+      // The socket bind failed or encountered a connection error in the test.
+      return false;
+    }
+    return true;
+  }
+
+  void SetUp() {
+    TLSServerRunner::start();
+    port_ = TLSServerRunner::port();
+  }
+
+ protected:
+  std::string port_;
+};
+
+TEST_F(TLSTransportsTests, test_call) {
+  // Create a transport and use a testing-only 'disableVerifyPeer' call.
+  // This allows our client to complete TLS without verifying the fake
+  // commonName or fake CA used by the testing server.
+  auto t = std::make_shared<TLSTransport>();
+  t->disableVerifyPeer();
+
+  // Create a request using a TLSTransport and JSONSerializer.
+  auto url = "https://localhost:" + port_;
+  auto r = Request<TLSTransport, JSONSerializer>(url, t);
+
+  // Use the 'call' method on the request without any input parameters.
+  // This will use a GET for the URI given in the Request constructor.
+  Status status;
+  ASSERT_NO_THROW(status = r.call());
+  if (verify(status)) {
+    pt::ptree recv;
+    status = r.getResponse(recv);
+    EXPECT_TRUE(status.ok());
+  }
+}
+
+TEST_F(TLSTransportsTests, test_call_with_params) {
+  // Again, use a fake server/CA/commonName certificate.
+  auto t = std::make_shared<TLSTransport>();
+  t->disableVerifyPeer();
```

Comment:
The cpp-netlib-enabling facility is NOT in stable. We are now maintaining a patch to enable unit testing.